sksports/model/heart_rate.py

```diff
@@ -0,0 +1,164 @@
+    """Model to predict heart-rate from power data."""
+    power_corrected = power + power.cumsum() * hr_drift
+    hr_lin = power_corrected * hr_slope + hr_rest
+    hr_ss = np.minimum(hr_max, hr_lin)
+
+    diff_hr = np.pad(np.diff(power_corrected), pad_width=1, mode='constant')
+    rate = np.where((diff_hr <= 0), rate_decay, rate_growth)
+
+    hr_pred = np.ones(power.shape) * hr_rest
+    for curr_idx in enumerate(hr_pred.size):
+        hr_prev = hr_pred[curr_idx] if curr_idx > 0 else hr_rest
+        hr_pred[curr_idx] = (hr_prev +
+                             (hr_ss[curr_idx] - hr_prev) / rate[curr_idx])
+
+    return hr_pred
+
+
+def _model_residual(params, heart_rate, power):
+    """Residuals between prediction and ground-truth."""
+    return heart_rate - _model(power, *params)
+
+
+def _model_residual_least_square(params, heart_rate, power):
+    """Least-square error of the residuals."""
+    return np.sum(_model_residual(params, heart_rate, power) ** 2)
+
+
+class HeartRateRegressor(BaseEstimator, RegressorMixin):
+    """Estimate the heart-rate from power data.
+
+    The model used is based on the formulation in [1]_.
+
+    Parameters
+    ----------
+    solver : string, optional
+        The solver to estimate the model parameters. Can be either 'leastsq' to
+        apply Levenberg-Marquardt or one of 'method' from
+        :func:`scipy.optimize.minimize`.
+
+    hr_rest : float, default=75
+        Heart-rate frequency of the cyclist at rest.
+
+    hr_max : float, default=200
+        Heart-rate frequency of the cyclist at maximum.
+
+    hr_slope : float, default=0.30
+        Slope considered if the model is linear. FIXME
+
+    hr_drift : float, default=3e-5
+        Attenuation of the heart-rate other time. FIXME
+
+    rate_growth : float, default=24
+        Growth rate of the exponential when the power increases.
+
+    rate_decay : float, default=30
+        Decay rate of the exponential when the power decreases.
+
+    Attributes
+    ----------
+    hr_rest_ : float
+        Fitted heart-rate frequency of the cyclist at rest.
+
+    hr_max_ : float
+        Fitted heart-rate frequency of the cyclist at maximum.
+
+    hr_slope_ : float
+        Fitted slope considered if the model is linear. FIXME
+
+    hr_drift_ : float
+        Fitted attenuation of the heart-rate other time. FIXME
+
+    rate_growth_ : float
+        Fitted growth rate of the exponential when the power increases.
+
+    rate_decay_ : float
+        Fitted decay rate of the exponential when the power decreases.
+
+    References
+    ----------
+    .. [1] de Smet, Dimitri, et al. "Heart rate modelling as a potential
+       physical fitness assessment for runners and cyclists." Workshop at ECML
+       & PKDD. 2016.
+
+    """
+
+    def __init__(self, solver='Nelder-Mead', hr_rest=75, hr_max=200,
+                 hr_slope=0.30, hr_drift=3e-5, rate_growth=24, rate_decay=30):
+        self.solver = solver
+        self.hr_rest = hr_rest
```

Comment:
Minor: I'd like to propose renaming `hr_rest` to `hr_start` or `hr_0` because this model parameter is not related to the actual resting heart rate of the athlete and therefore might cause confusion.